src/DomainNetMap.cpp

```diff
@@ -0,0 +1,265 @@
+            throw Exception("Neural net file exceeds maximum size.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        std::string buf, err;
+        buf.reserve(length);
+
+        std::getline(file, buf, '\0');
+
+        const json11::Json nnet_json = json11::Json::parse(buf, err);
+
+        // make sure that there no unexpected keys in the json
+        std::vector<std::string> expected_keys = {"layers", "signal_inputs", "delta_inputs", "trace_outputs", "description"};
+        for (auto &key : nnet_json.object_items()) {
+            if (std::find(expected_keys.begin(), expected_keys.end(), key.first) == expected_keys.end()) {
+                throw Exception(
+                        "Unexpected key in neural net json: " + key.first + "\n",
+                        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+        }
+
+        if (! nnet_json["layers"].is_array()) {
+            throw Exception(
+                    "Neural net must contain valid json and must have "
+                    "a key \"layers\" whose value is an array.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if (nnet_json["signal_inputs"].array_items().size() == 0 
+            && nnet_json["delta_inputs"].array_items().size() == 0) {
+            throw Exception(
+                    "Neural net json must contain at least one of "
+                    "\"signal_inputs\" and \"delta_inputs\" whose "
+                    "value is a non-empty array.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        
+        std::vector<std::shared_ptr<DenseLayer>> layers;
+        for (std::size_t layer_idx = 0; layer_idx < nnet_json["layers"].array_items().size(); ++layer_idx) {
+            layers.push_back(json_to_DenseLayer(nnet_json["layers"][layer_idx]));
+        }
+
+	if (layers.size() == 0) {
+            throw Exception(
+                    "Layers must be non-empty.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        m_neural_net = m_nn_factory->createLocalNeuralNet(layers);
+
+	if (nnet_json["signal_inputs"].array_items().size()
+                + nnet_json["delta_inputs"].array_items().size()
+                != m_neural_net->get_input_dim()) {
+            throw Exception(
+                    "Neural net input dimension must match the number of "
+                    "signal and delta inputs.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+	}
+
+        if (nnet_json["trace_outputs"].array_items().size()
+                != m_neural_net->get_output_dim()) {
+            throw Exception(
+                    "Neural net output dimension must match the number of "
+                    "trace outputs.\n",
+                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        for (std::size_t i=0; i<nnet_json["signal_inputs"].array_items().size(); i++) {
```

Comment:
```suggestion
        for (const auto &signal : nnet_json["signal_inputs"]) {
```

Modify lines below accordingly.